### PR TITLE
Adding UCC build

### PR DESCRIPTION
**Pull request by nirandaperera**

### CODE REVIEW

**Comment by vibhatha** [comment]
Thanks for this PR @nirandaperera 
I will look into this one.

File: cpp/src/examples/CMakeLists.txt [hunk 1]
```diff
@@ -42,7 +42,20 @@ macro(cylon_add_exe EXENAME)
...
+#        if (CYLON_UCC)
+#            target_link_libraries(${EXENAME} ucc)
+#            target_link_libraries(${EXENAME} ucc /home/niranda/git/ucc/install/lib/ucc/libucc_cl_basic.so
+#                    /home/niranda/git/ucc/install/lib/ucc/libucc_cl_hier.so
+#                    /home/niranda/git/ucc/install/lib/ucc/libucc_mc_cpu.so
+#                    /home/niranda/git/ucc/install/lib/ucc/libucc_tl_ucp.so)
+##            target_include_directories(${EXENAME} INTERFACE ${UCC_INCLUDE_DIRS})
+#            include_directories(${UCC_INCLUDE_DIRS})
+#        endif (CYLON_UCC)
+#    endif (CYLON_UCX)
```
Comment (vibhatha) [inline cpp/src/examples/CMakeLists.txt:58]: Do we need these?

**Comment by nirandaperera** [comment]
I believe there's an issue in the `ucc_example.cpp` linking. I was able to compile and run the example based on this 
https://github.com/openucx/ucc/wiki/UCC-Allreduce-example

But I still couldn't get it properly linked inside cylon examples unfortunately. :-(

File: cpp/src/examples/ucc_example.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+#define STR(x) #x
+#define UCC_CHECK(_call)                                            \
+    if (UCC_OK != (_call)) {                                        \
+        fprintf(stderr, "*** UCC TEST FAIL: %s\n", STR(_call));     \
+        MPI_Abort(MPI_COMM_WORLD, -1);                              \
+    }
+
+static ucc_status_t oob_allgather(void *sbuf, void *rbuf, size_t msglen,
+                                  void *coll_info, void **req) {
```
Comment (vibhatha) [inline cpp/src/examples/ucc_example.cpp:31]: Are we going to C style for method naming or should we go for Camel case?

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -22,7 +22,7 @@
 # For more information about CMake, see http://www.cmake.org             #
 ##########################################################################
 cmake_minimum_required(VERSION 3.17 FATAL_ERROR)
-set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+#set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
```
Comment (vibhatha) [inline cpp/CMakeLists.txt:25]: Do we need this?

File: cpp/src/examples/ucc_example.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+ *  limitations under the License.
+ */
+
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <ucc/api/ucc.h>
+#include <iostream>
```
Comment (vibhatha) [inline cpp/src/examples/ucc_example.cpp:21]: order of imports

**Review by vibhatha — CHANGES_REQUESTED** [review]
This is just a quick peek on the PR, I will review again. 

Just a thought, should we add some docs on how to download, compile, which versions we are supporting?

**Comment by vibhatha** [comment]
> I believe there's an issue in the `ucc_example.cpp` linking. I was able to compile and run the example based on this https://github.com/openucx/ucc/wiki/UCC-Allreduce-example
> 
> But I still couldn't get it properly linked inside cylon examples unfortunately. :-(

Let's see, I will also look into this.

File: cpp/CMakeLists.txt (hunk 3 above)
Comment (nirandaperera) [reply cpp/CMakeLists.txt:25]: may be this was the problem. I added this to run examples without having to update `LD_LIBRARY_PATH`. It seems like we'll have to figure out an alternative to that

Comment (nirandaperera) [reply cpp/CMakeLists.txt]: I figured out the issue. It turned out that I had a built ucc with a different ucx version than the one in my conda env.

File: cpp/src/examples/ucc_example.cpp (hunk 2 above)
Comment (nirandaperera) [reply cpp/src/examples/ucc_example.cpp:34]: This is just an example. We can remove this once UCC is properly integrated.

**Pull request merged**